common/src/chain/transaction/signature/inputsig/authorize_pubkeyhash_spend.rs

```diff
@@ -77,4 +77,223 @@ pub fn sign_address_spending(
+                sighashtype::SigHashType,
+                signature_hash,
+            },
+            Destination, Transaction, TransactionCreationError, TxInput, TxOutput,
+        },
+        primitives::{Amount, Id},
+    };
+
+    use super::*;
+
+    fn generate_unsigned_tx(
+        outpoint_dest: Destination,
+    ) -> Result<Transaction, TransactionCreationError> {
+        let tx = Transaction::new(
+            0,
+            vec![TxInput::new(
+                Id::<Transaction>::new(&H256::zero()).into(),
+                0,
+                InputWitness::NoSignature(None),
+            )],
+            vec![TxOutput::new(Amount::from_atoms(100), outpoint_dest)],
+            0,
+        )?;
+        Ok(tx)
+    }
+
+    fn prepare_data_for_verify(
+        sighash_type: SigHashType,
+    ) -> (PublicKeyHash, AuthorizedPublicKeyHashSpend, H256) {
+        const INPUT_NUM: usize = 0;
+        let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let pubkey_hash = PublicKeyHash::from(&public_key);
+        let outpoint_dest = Destination::Address(pubkey_hash);
+        let tx = generate_unsigned_tx(outpoint_dest.clone()).unwrap();
+        let witness = StandardInputSignature::produce_signature_for_input(
+            &private_key,
+            sighash_type,
+            outpoint_dest,
+            &tx,
+            INPUT_NUM,
+        )
+        .unwrap();
+        let spender_signature =
+            AuthorizedPublicKeyHashSpend::from_data(witness.get_raw_signature()).unwrap();
+        let sighash = signature_hash(witness.sighash_type(), &tx, INPUT_NUM).unwrap();
+        (pubkey_hash, spender_signature, sighash)
+    }
+
+    fn prepare_data_for_wrong_destination(sighash_type: SigHashType) -> StandardInputSignature {
+        const INPUT_NUM: usize = 0;
+        let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let outpoint_dest = Destination::PublicKey(public_key);
+        let tx = generate_unsigned_tx(outpoint_dest.clone()).unwrap();
+        StandardInputSignature::produce_signature_for_input(
+            &private_key,
+            sighash_type,
+            outpoint_dest,
+            &tx,
+            INPUT_NUM,
+        )
+        .unwrap()
+    }
+
+    #[test]
+    fn test_wrong_destination() {
+        use std::mem::discriminant;
+        // Destination = Address, but we try to use here AuthorizedPublicKeyHashSpend
+        let sighash_type = SigHashType::try_from(SigHashType::ALL).unwrap();
+        let witness = prepare_data_for_wrong_destination(sighash_type);
+        // Compare Err(TransactionSigError) without inner message
+        assert_eq!(
+            discriminant(&AuthorizedPublicKeyHashSpend::from_data(
+                witness.get_raw_signature()
+            )),
+            discriminant(&Err(TransactionSigError::AddressAuthDecodingFailed(
+                String::new()
+            )))
+        );
+        let sighash_type =
```

Comment:
When you test the same thing multiple times in one test, it's extremely helpful for the reader and for safety to isolate these tests. Usually this is done by having multiple unit tests, but you're not required to do that specifically. An easier solution would be to just wrap every case in its own scope with `{}`. So here, start with `{` at line 153 and close with `}` at line 161.

The reason for this is two fold:

1. Like I said before, easier for the reader to understand the scope of the test.
2. Safer, since variables can't collide and shadowing of variables won't be necessary. Carrying the state of a variable through different tests and accidently using it is a dangerous things. So having a separate scope for tests is a good idea to prevent things like that from happening.